pkg/util/objectpool/withlock/objectpool.go

```diff
@@ -0,0 +1,179 @@
+		lock         sync.Mutex                  // lock for sync
+		checkWhenGet bool                        // whether to health check when get IPoolObject
+		checkWhenPut bool                        // whether to health check when put IPoolObject
+	}
+
+	// Spec Pool's spec
+	Spec struct {
+		initSize     int32                       // initial size
+		maxSize      int32                       // max size
+		new          func() (IPoolObject, error) // create a new object
+		checkWhenGet bool                        // whether to health check when get IPoolObject
+		checkWhenPut bool                        // whether to health check when put IPoolObject
+	}
+)
+
+// NewSimplePool returns a new pool
+func NewSimplePool(initSize, maxSize int32, new func() (IPoolObject, error)) *Pool {
+	return NewPoolWithSpec(Spec{
+		initSize:     initSize,
+		maxSize:      maxSize,
+		new:          new,
+		checkWhenGet: true,
+		checkWhenPut: true,
+	})
+}
+
+// NewPoolWithSpec returns a new pool
+func NewPoolWithSpec(spec Spec) *Pool {
+	p := &Pool{
+		initSize:     spec.initSize,
+		maxSize:      spec.maxSize,
+		new:          spec.new,
+		checkWhenPut: spec.checkWhenPut,
+		checkWhenGet: spec.checkWhenGet,
+		store:        make(chan IPoolObject, spec.maxSize),
+	}
+	if err := p.init(); err != nil {
+		logger.Errorf("new pool failed %v", err)
+		return nil
+	}
+	return p
+}
+
+// Validate validate
+func (s *Spec) Validate() error {
+	if s.initSize > s.maxSize {
+		s.maxSize = s.initSize
+	}
+	if s.maxSize <= 0 {
+		return fmt.Errorf("pool max size must be positive")
+	}
+	if s.initSize < 0 {
+		return fmt.Errorf("pool init size must greate than or equals 0")
+	}
+
+	return nil
+}
+
+// initializes the pool with initSize of objects
+func (p *Pool) init() error {
+	for i := 0; i < int(p.initSize); i++ {
+		iPoolObject, err := p.new()
+		if err != nil || !iPoolObject.HealthCheck() {
+			logger.Errorf("create pool object failed or object of pool is not healthy when init pool:, err %v", err)
+			continue
+		}
+		atomic.AddInt32(&p.size, 1)
+		p.store <- iPoolObject
+	}
+	return nil
+}
+
+// Get returns an object from the pool,
+// if no free object, it will create a new one if the pool is not full
+// if the pool is full, it will block until an object is returned to the pool
+func (p *Pool) Get(ctx context.Context) (IPoolObject, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case poolObject := <-p.store:
+			if p.checkWhenGet && !poolObject.HealthCheck() {
+				p.destroyIPoolObject(poolObject)
+				if poolObject = p.createIPoolObject(); poolObject != nil {
+					return poolObject, nil
+				}
+				continue
+			}
+			return poolObject, nil
+		default:
+			if iPoolObject := p.createIPoolObject(); iPoolObject != nil {
+				return iPoolObject, nil
+			}
+		}
+	}
+}
+
+func (p *Pool) createIPoolObject() IPoolObject {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	for i := 0; i < int(p.maxSize-p.size); i++ {
```

Comment:
why use a loop here?

Reply:
If the `new` function returns err, I hope to retry in the `createIPoolObject` function instead of back to the `Get` function and waiting for the next time. The latter could have a higher probability of context switching. Of course, on the other hand, rapid retrying will still get an err, here I think both have advantages and disadvantages

Reply:
as you mentioned "rapid retrying will still get an err", I propose not using the loop.